Predictive back route transitions by default

justinmc opened the pull request:
This PR turns on predictive back route transitions by default on supported Android devices.

With https://github.com/flutter/flutter/pull/154718, the default (PredictiveBackPageTransitionsBuilder) is the [shared element transition](https://developer.android.com/design/ui/mobile/guides/patterns/predictive-back#shared-element-transition). The [full screen transition](https://developer.android.com/design/ui/mobile/guides/patterns/predictive-back#full-screen-surfaces) is also available by using PredictiveBackFullScreenPageTransitionsBuilder.

Original PR: https://github.com/flutter/flutter/pull/146788

Depends on: https://github.com/flutter/flutter/pull/154718

When this lands in stable, the docs should be updated: https://docs.flutter.dev/platform-integration/android/predictive-back

justinmc committed:
Predictive back is the default transition

Where not supported, Zoom will be used.

justinmc committed:
Test

justinmc committed:
Analyzer

justinmc committed:
Merge branch 'master' into predictive-back-by-default-2

justinmc committed:
Merge branch 'master' into predictive-back-by-default-2

justinmc committed:
dart format

justinmc committed:
Fix test

justinmc committed:
Merge branch 'master' into predictive-back-by-default-2

justinmc committed:
Fix expansiontile test

justinmc committed:
Fix flexiblespacebartest

justinmc committed:
Fix page_test.dart

justinmc committed:
Fix snack bar test

justinmc committed:
Fix text field test

justinmc committed:
Analyzer (and unnecessary extra param)

justinmc committed:
Fix icon_button_test

justinmc committed:
Fix bottom app bar test

justinmc committed:
Fix page_transitions_theme_test

And get rid of all instances of _PredictiveBackPageTransition (old
name).

justinmc committed:
Analyzer

chunhtai commented on packages/flutter/lib/src/material/page_transitions_theme.dart:
```diff
@@ -1091,7 +1092,7 @@ class PageTransitionsTheme with Diagnosticable {
   static const Map<TargetPlatform, PageTransitionsBuilder> _defaultBuilders =
       <TargetPlatform, PageTransitionsBuilder>{
-        TargetPlatform.android: ZoomPageTransitionsBuilder(),
+        TargetPlatform.android: PredictiveBackPageTransitionsBuilder(),
```
will this be a breaking change? I am not too familiar with the process of changing page transition. so maybe @QuncCccccc can chime in.

An alternative is some how expose the predicitiveback enabled setting to framework side and conditionally return PredictiveBackPageTransitionsBuilder or ZoomPageTransitionsBuilder.

justinmc replied:
That's actually what it does internally:

https://github.com/flutter/flutter/blob/8d2c9782e334e5f14066b586a6e6cac916b17608/packages/flutter/lib/src/material/predictive_back_page_transitions_builder.dart#L73-L93

So this PR moves us from Zoom to FadeForwards by default, and when doing a predictive back gesture, it will use the _PredictiveBackSharedElement transition by default.

Can you verify that we're good to make FadeForwards the default @QuncCccccc?

benthillerkus replied:
And it will always use that fallback aslong as you don't opt in in the AndroidManifest.xml, right?

justinmc replied:
Yes that's right.

justinmc committed:
Fix divider example test

justinmc committed:
Fix hero example test using hardcoded transition duration

justinmc committed:
Fix listener example test

justinmc committed:
Fix mouse region example test

justinmc committed:
Fix align transition example test

justinmc committed:
Fix slide transition example test

justinmc committed:
Fix fade transition example

justinmc committed:
Formatting

justinmc committed:
Fix slivers evil test

justinmc committed:
Fix routes test for hardcoded transition duration

justinmc committed:
Fix heroes test hardcoded transition duration

justinmc committed:
Clean up other heroes test hardcoded transition duration values

justinmc committed:
Fix navigator test for hardcoded transition duration

justinmc committed:
Fix hero 1 example test due to hardcoded transition duration

justinmc committed:
Merge branch 'master' into predictive-back-by-default-2

justinmc committed:
Format

justinmc committed:
Fix hardcoded transition duration

justinmc commented:
Looks like there were customer test failures on the previous commit ([example in zulip](https://github.com/zulip/zulip-flutter/blob/main/test/widgets/home_test.dart#L373)). I think they may be due to the fact that the transition time is increasing.

justinmc committed:
Formatting

justinmc committed:
TEST: Using default page transition duration.

If the duration is the same, do customer tests pass?

justinmc commented:
@gnprice It looks like some zulip tests fail if I change the default page transition duration. Example: https://ci.chromium.org/ui/p/flutter/builders/try/Linux%20customer_testing/100474/overview

Previously the Zoom transition used:

https://github.com/flutter/flutter/blob/11bf180af0bd2b03ccbec636249250dbd10384f9/packages/flutter/lib/src/material/page_transitions_theme.dart#L627

Now FadeForwards will be the default and it uses:

https://github.com/flutter/flutter/blob/11bf180af0bd2b03ccbec636249250dbd10384f9/packages/flutter/lib/src/material/page_transitions_theme.dart#L759-L762

(Though there has been some controversy over whether that's the correct duration or not, see https://github.com/flutter/flutter/pull/168425#issuecomment-2856242674).

From Zulip's point of view, would it make sense to use something programmatic like kTransitionMilliseconds or transitionDuration instead of hardcoding it in your code, and would that fix the problem in a robust way?

Or alternatively, maybe Flutter needs to rethink adjusting this value if it's going to break customers and we're not 100% confident in it.

TheLastFlame commented:
@justinmc, actually, that transition looks much slower than what is used in Android. Also, I have not been able to find any of the specs that QuncCccccccc mentions. I have not independently studied the Android source code, but I have run Deep Research a couple times for this task. I was getting a value of 450ms output.

justinmc commented:
+1 that it feels a lot faster on native Android. I've confirmed the 800ms duration with someone on the Material team, but they are also using springs from Material 3 Expressive. I think the result is it looks very fast during the main animation but has a very slow ending that lasts until 800ms.

Flutter could fudge it for now by using 450ms, but it looks like that might still break Zulip's tests.

gnprice commented:
(/cc @chrisbobbe for looking in more detail, since I'm OOO this week)

We definitely have some tests in Zulip that wait a fixed number of ms when what they really want is "until the page transition finishes", and the real point of the test doesn't care whether that takes 300ms or 800ms or some other value. (I'd expect Zulip to be representative of a lot of apps out there in having some tests of this kind.) So where that's the case, the test would be cleaner if there were some value it could refer to instead.

It might be tricky to specify the value as a constant, though — different transitions may take different durations, right? So maybe the API that's needed is some way to look up the duration (or the remaining duration?) of the page transition that's currently in progress, or perhaps to get a Future that will complete when the transition finishes.

justinmc commented:
Got it, thanks for the post while you're OOO. I'll see if I can find a way to expose the duration, which will also cover cases where people want to wait for half the duration, etc.

justinmc commented:
The Google failures are strangely all very slight opacity changes in a few different things. This is with my test that keeps the transition duration at 300.

justinmc committed:
Revert "TEST: Using default page transition duration."

This reverts commit 7f80cd22fcac461bcb744ea1ed44b1b78742626e.

chrisbobbe commented:
Chris from Zulip here 👋. To wait through a pop-route animation, how about tracking the last-popped route and using its `reverseTransitionDuration`:

```dart
      Route<dynamic>? lastPopped;
      final testNavObserver = TestNavigatorObserver()
        ..onPopped = (route, _) => lastPopped = route;
```

```dart
      // TODO could do some checks on lastPopped
      await tester.pump((lastPopped as TransitionRoute).reverseTransitionDuration);
      await tester.pump(Duration(milliseconds: 1)); // fails without this; don't know why
```


<details>
<summary>`TestNavigatorObserver`</summary>

```dart
import 'package:flutter/widgets.dart';

// Inspired by test code in the Flutter tree:
//   https://github.com/flutter/flutter/blob/53082f65b/packages/flutter/test/widgets/observer_tester.dart
//   https://github.com/flutter/flutter/blob/53082f65b/packages/flutter/test/widgets/navigator_test.dart

/// A trivial observer for testing the navigator.
class TestNavigatorObserver extends NavigatorObserver {
  void Function(Route<dynamic> route, Route<dynamic>? previousRoute)? onPushed;
  void Function(Route<dynamic> route, Route<dynamic>? previousRoute)? onPopped;
  void Function(Route<dynamic> route, Route<dynamic>? previousRoute)? onRemoved;
  void Function(Route<dynamic>? route, Route<dynamic>? previousRoute)? onReplaced;
  void Function(Route<dynamic> route, Route<dynamic>? previousRoute)? onStartUserGesture;
  void Function()? onStopUserGesture;

  @override
  void didPush(Route<dynamic> route, Route<dynamic>? previousRoute) {
    onPushed?.call(route, previousRoute);
  }

  @override
  void didPop(Route<dynamic> route, Route<dynamic>? previousRoute) {
    onPopped?.call(route, previousRoute);
  }

  @override
  void didRemove(Route<dynamic> route, Route<dynamic>? previousRoute) {
    onRemoved?.call(route, previousRoute);
  }

  @override
  void didReplace({ Route<dynamic>? oldRoute, Route<dynamic>? newRoute }) {
    onReplaced?.call(newRoute, oldRoute);
  }

  @override
  void didStartUserGesture(Route<dynamic> route, Route<dynamic>? previousRoute) {
    onStartUserGesture?.call(route, previousRoute);
  }

  @override
  void didStopUserGesture() {
    onStopUserGesture?.call();
  }
}
```

</details>

justinmc committed:
Test of possible approach to avoid hardcoding.

By hardcoding I mean both hardcoding the duration and hardcoding the
page transtion (using
FadeForwardsPageTransitionBuilder.kTransitionMilliseconds assumes that
that is the default transition).

justinmc committed:
Merge branch 'master' into predictive-back-by-default-2

justinmc committed:
Fix test descendent path

justinmc commented:
Thank you for quickly unblocking this @chrisbobbe! I will update this PR to use something like TestNavigatorObserver in the framework tests where transition duration is needed, and see what @gnprice says next week.

justinmc committed:
Merge branch 'master' into predictive-back-by-default-2

justinmc committed:
TransitionDurationObserver.

justinmc committed:
Let's keep the kTransitionMilliseconds constant.

The predictive back transition needs to match Fade Forwards' duration.
If Fade Forwards changed, then predictive back should change too.

justinmc committed:
Docs analyzer

justinmc committed:
Merge branch 'master' into predictive-back-by-default-2

justinmc committed:
Analyzer

justinmc committed:
I was missing some specifics about which transition duration is used.

E.g. when popping, the duration of the incoming route is used. In
general, it uses the same duration forwards and backwards.

justinmc commented:
Alright I've created [TransitionDurationObserver](https://github.com/flutter/flutter/pull/165832/files#diff-df963347641565870466157a051a509203fa31315ecdf6cf9e12282dfae60e32) to be used to get the transition duration in tests. When you use it, your test is not dependent on which transition is the default or on the duration of that transition.

The best commit to look at to see the changes required of the framework to use this is https://github.com/flutter/flutter/pull/165832/commits/2a7907173c6e8bd1d62bbbc39c98d51a04913a55.

Example:

```dart
testWidgets('example', (WidgetTester tester) async {
  final TransitionDurationObserver observer = TransitionDurationObserver();

  await tester.pumpWidget(
    MaterialApp(
      navigatorObservers: <NavigatorObserver>[observer],
      onGenerateRoute: (RouteSettings settings) { ... },
    ),
  );

  expect(find.text('Page 1'), findsOneWidget);
  expect(find.text('Page 2'), findsNothing);

  await tester.tap(find.text('Next'));

  await tester.pump();
  await tester.pump(observer.transitionDuration + const Duration(milliseconds: 1));

  expect(find.text('Page 1'), findsNothing);
  expect(find.text('Page 2'), findsOneWidget);
});
```

justinmc committed:
Analyzer

justinmc commented:
@chrisbobbe This still seems to be failing on Zulip tests. Is this actually still causing a failure or is something out of date with CI? The line numbering seems to point to an older revision if I'm not mistaken.

<details>

<summary>Test failure output, points to getRouteOf on line 373</summary>

https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8711775929764274657/+/u/run_test.dart_for_customer_testing_shard_and_subshard_None/stdout

```
| 01:03 +1714 ~15: /b/s/w/ir/x/t/flutter_customer_testing.zulip.DFHCDG/tests/test/widgets/home_test.dart: _LoadingPlaceholderPage while loading, choosing an account disallows going back
| ══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
| The following StateError was thrown running a test:
| Bad state: Too many elements
|
| When the exception was thrown, this was the stack:
| #0      Iterable.single (dart:core/iterable.dart:696:24)
| #1      WidgetController.element (package:flutter_test/src/controller.dart:864:30)
| #2      main.getRouteOf (file:///b/s/w/ir/x/t/flutter_customer_testing.zulip.DFHCDG/tests/test/widgets/home_test.dart:60:26)
| #3      main.<anonymous closure>.<anonymous closure> (file:///b/s/w/ir/x/t/flutter_customer_testing.zulip.DFHCDG/tests/test/widgets/home_test.dart:373:13)
| <asynchronous suspension>
| #4      testWidgets.<anonymous closure>.<anonymous closure> (package:flutter_test/src/widget_tester.dart:193:15)
| <asynchronous suspension>
| #5      TestWidgetsFlutterBinding._runTestBody (package:flutter_test/src/binding.dart:1064:5)
| <asynchronous suspension>
| <asynchronous suspension>
| (elided one frame from package:stack_trace)
```

</details>

But those seem to be the old line numbers before you [bumped](https://github.com/flutter/tests/pull/460) the customer tests.

https://github.com/zulip/zulip-flutter/pull/1547/files#diff-4b37bb0d549c9a4ef3f098301b812ea9e864e3a2163d6fd10b2bed26c6173ba5L373

![Screenshot from 2025-06-18 14-40-52](https://github.com/user-attachments/assets/1dc16d64-da8e-45e1-ac39-dc545033ade2)

justinmc commented:
Ah I missed that PR, thanks.

Splitting out TransitionDurationObserver sounds like a good idea. Zulip is the only customer broken in customer_testing, but there are several Google test failures too, and probably many other customers not in customer_testing.

dkwingsmt commented:
Is this PR blocked by https://github.com/flutter/flutter/pull/171109? (from triage)

justinmc committed:
Merge branch 'master' into predictive-back-by-default-2

justinmc committed:
Formatting

justinmc committed:
Fix stepper test

Once again a selector matched something in the page transition and
making it more specific fixed it.

gnprice commented:
Didn't see that latest comment until today, sorry. (*)

Investigating that test, it looks like it's tied to the specifics of how the page transition animation works. The longer duration doesn't break this one; but it does assume that in a "back" transition, neither the old nor new page is getting translated around. That works in the current default transition, but naturally gets broken by the new one.

Probably the right answer is we should skip the test while this lands, and potentially follow up by reworking the test so that it can do its job with the new transitions.

(The thing the test is meant to do is to check we don't have a spurious Hero animation between two pages even though they have an image in common — a regression test for https://github.com/zulip/zulip-flutter/issues/930 . To do this it finds the image element and its rect while on the first page; navigates to the second page, finds the image element there and its rect; then hits the back button, and throughout the transition it repeatedly checks that both image elements remain at the same respective places. The failure is because with the new transition, the old page's image starts out 200px to the left of where it was when that page was on top. The next step would also fail, because the new page and its image have started to animate away.)

(*) Any time there's a Zulip customer test failure that might be blocking something, please feel free to send us a direct email, or ping by Discord — GitHub notifications are a firehose where I've always found it hard to avoid missing things.

gnprice commented:
Sent a PR to skip that test: https://github.com/zulip/zulip-flutter/pull/1741

Once that lands, the pin will just need to be bumped to take that change (and the customer_testing pin in this repo bumped to take that change in turn).

justinmc commented:
Thank you @gnprice! I have the same firehose problem with GitHub notifications. Looks like your PR already landed, so I'll be bumping the SHAs.

justinmc committed:
Merge branch 'master' into predictive-back-by-default-2

justinmc commented:
The previous merge includes the updated customer tests, so I'd expect the customer tests to go green here (still work to do on the Google tests though).

justinmc commented:
@gnprice @chrisbobbe Looks like a new Zulip failure. I'm winding down on Friday but will check in with you guys on Monday on Discord if needed.

<details>

<summary>Failing test output</summary>

```
| 01:27 +2132 ~8: /b/s/w/ir/x/t/flutter_customer_testing.zulip.DAWBRS/tests/test/widgets/message_list_test.dart: OutboxMessageWithPossibleSender hidden -> failed, tapping does nothing if compose box is not offered
| ══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
| The following assertion was thrown running a test:
| Finder specifies a widget that would not receive pointer events.
| A call to tap() with finder "Found 1 widget with icon "IconData(U+0F128)": [
|   Icon(IconData(U+0F128), color: Color(alpha: 0.5000, red: 0.3804, green: 0.3490, blue: 0.8824,
| colorSpace: ColorSpace.sRGB), dependencies: [Directionality, IconTheme]),
| ]" derived an Offset (Offset(970.0, 571.3)) that would not hit test on the specified widget.
| Maybe the widget is actually off-screen, or another widget is obscuring it, or the widget cannot
| receive pointer events.
| Indeed, Offset(970.0, 571.3) is outside the bounds of the root of the render tree, Size(800.0,
| 600.0).
| The finder corresponds to this RenderBox: RenderSemanticsAnnotations#29ab9 relayoutBoundary=up30 NEEDS-PAINT(creator: Semantics ← Icon ← Semantics ← Align ← Padding ← Listener ← RawGestureDetector ← GestureDetector ← Semantics ← DefaultSelectionStyle ← Builder ← MouseRegion ← ⋯, parentData: <none> (can use size), constraints: BoxConstraints(0.0<=w<=28.0, 0.0<=h<=28.0), size: Size(24.0, 24.0))
| The hit test result at that offset is:
| HitTestResult(HitTestEntry<HitTestTarget>#760f4(_ReusableRenderView#f0ce3),
| HitTestEntry<HitTestTarget>#4fdae(<AutomatedTestWidgetsFlutterBinding>))
| If you expected this target not to be able to receive pointer events, pass "warnIfMissed: false" to
| "tap()".
| To make this error into a non-fatal warning, set WidgetController.hitTestWarningShouldBeFatal to
| false.
|
| When the exception was thrown, this was the stack:
| #0      WidgetController._getElementPoint (package:flutter_test/src/controller.dart:2045:11)
| #1      WidgetController.getCenter (package:flutter_test/src/controller.dart:1861:12)
| #2      WidgetController.tap (package:flutter_test/src/controller.dart:1041:7)
| #3      main.<anonymous closure>.sendMessageAndFail (file:///b/s/w/ir/x/t/flutter_customer_testing.zulip.DAWBRS/tests/test/widgets/message_list_test.dart:2014:20)
| <asynchronous suspension>
| #4      main.<anonymous closure>.<anonymous closure> (file:///b/s/w/ir/x/t/flutter_customer_testing.zulip.DAWBRS/tests/test/widgets/message_list_test.dart:2104:7)
| <asynchronous suspension>
| #5      testWidgets.<anonymous closure>.<anonymous closure> (package:flutter_test/src/widget_tester.dart:192:15)
| <asynchronous suspension>
| #6      TestWidgetsFlutterBinding._runTestBody (package:flutter_test/src/binding.dart:1059:5)
| <asynchronous suspension>
| <asynchronous suspension>
| (elided one frame from package:stack_trace)
|
| The test description was:
|   hidden -> failed, tapping does nothing if compose box is not offered
| ════════════════════════════════════════════════════════════════════════════════════════════════════
|
| 01:27 +2132 ~8 -1: /b/s/w/ir/x/t/flutter_customer_testing.zulip.DAWBRS/tests/test/widgets/message_list_test.dart: OutboxMessageWithPossibleSender hidden -> failed, tapping does nothing if compose box is not offered [E]
|   Test failed. See exception logs above.
|   The test description was: hidden -> failed, tapping does nothing if compose box is not offered
```

</details>

justinmc committed:
Merge branch 'master' into predictive-back-by-default-2

justinmc commented:
That did it to get the Zulip tests passing, thanks @gnprice @chrisbobbe.

Now working on the Google test failures.

justinmc committed:
Merge branch 'master' into predictive-back-by-default-2

justinmc committed:
Update the transition duration

justinmc commented:
I've changed the transition duration (details in https://github.com/flutter/flutter/issues/173252). The reason why I'm doing it here is that I worry about needing to change this duration again later due to all of the breakages we've been dealing with (even though our fixes should be making these tests more robust). So I'm going to change both the default transition and its duration in one go.

FYI @QuncCccccc, when you are back at work it would be great to get your approval on the value I'm using here. You can make the change from commit https://github.com/flutter/flutter/pull/165832/commits/1b7f841a5a4ef1c5e28a028f330fb86b60e12360 locally on your machine and run an app with Fade Forwards transitions like the one below.

<details>

<summary>Transition demo app</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MainApp());
}

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        brightness: Brightness.light,
        pageTransitionsTheme: const PageTransitionsTheme(
          builders: {
            TargetPlatform.android: PredictiveBackPageTransitionsBuilder(),
            TargetPlatform.iOS: CupertinoPageTransitionsBuilder(),
            TargetPlatform.macOS: CupertinoPageTransitionsBuilder(),
          },
        ),
      ),
      home: const FirstScreen(),
    );
  }
}

class FirstScreen extends StatelessWidget {
  const FirstScreen({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('First Screen')),
      body: Center(
        child: Column(
          children: <Widget>[
            ElevatedButton(
              onPressed: () {
                Navigator.of(
                  context,
                ).push(MaterialPageRoute(builder: (_) => const SecondScreen()));
              },
              child: const Text('Go to Second Screen'),
            ),
          ],
        ),
      ),
    );
  }
}

class SecondScreen extends StatelessWidget {
  const SecondScreen({super.key});

  @override
  Widget build(BuildContext context) {
    return Theme(
      //data: ThemeData(brightness: Brightness.dark),
      data: ThemeData(brightness: Brightness.light),
      child: Scaffold(
        appBar: AppBar(title: const Text('Second Screen')),
        body: Center(
          child: PopScope(
            canPop: true,
            onPopInvoked: (bool didPop) {
              print('justin popinvoked $didPop');
            },
            child: const Text('Hello, Predictive back!'),
          ),
        ),
      ),
    );
  }
}
```

</details>

![450ms](https://github.com/user-attachments/assets/594d9afe-5fe7-4acc-83c2-a62d97a7285a)

justinmc commented:
We seem to have a fix for the failing Google tests thanks to @victorsanni. On Monday I will confirm the transition duration and the fix and then move forward with getting this merged.

QuncCccccc commented on packages/flutter/lib/src/material/page_transitions_theme.dart:
```diff
@@ -758,7 +759,12 @@ class FadeForwardsPageTransitionsBuilder extends PageTransitionsBuilder {
+  /// the actual value used by native Android, which is 800ms, because native
+  /// Android is using Material 3 Expressive springs that are not currently
+  /// supported by Flutter. So for now at least, this is an approximation.
+  static const int kTransitionMilliseconds = 450;
```
I just compared the native behavior and our transition with 450 ms transition duration. I think with the updated value, it looks a lot better than the original! 

https://github.com/user-attachments/assets/a1d0edd9-7e78-4305-a59e-045daefbd429

QuncCccccc reviewed (approved):
LGTM on the default duration change:)! Thanks for the improvement!

matanlurey commented:
reason for revert: Breaks `Linux_pixel_7pro embedded_android_views_integration_test`:

- https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_pixel_7pro%20embedded_android_views_integration_test/8918/overview
- https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_pixel_7pro%20embedded_android_views_integration_test/8917/overview

```txt
[2025-08-14 16:01:17.600761] [STDOUT] stdout: [   +1 ms]   Expected: ''
[2025-08-14 16:01:17.600828] [STDOUT] stdout: [        ]     Actual: '[pointerCoord 0] x (expected: 473.0 actual: 472.990234375) [pointerCoord 1] x (expected: 370.0 actual: 369.990234375) [pointerCoord 2] x (expected: 809.0 actual: 808.990234375) , [pointerCoord 0] x (expected: 471.0 actual: 470.990234375) [pointerCoord 1] x (expected: 369.0 actual: 368.990234375) [pointerCoord 2] x (expected: 809.0 actual: 808.990234375) , [pointerCoord 0] x (expected: 470.0 actual: 468.990234375) [pointerCoord 1] x (expected: 368.0 actual: 366.990234375) [pointerCoord 2] x (expected: 809.0 actual: 807.990234375) , [pointerCoord 0] x (expected: 469.0 actual: 468.9052734375) [pointerCoord 1] x (expected: 367.0 actual: 366.9052734375) [pointerCoord 2] x (expected: 809.0 actual: 808.90625) , [pointerCoord 0] x (expected: 468.0 actual: 467.9052734375) [pointerCoord 1] x (expected: 366.0 actual: 365.9052734375) [pointerCoord 2] x (expected: 809.0 actual: 808.90625) , [pointerCoord 0] x (expected: 467.0 actual: 466.7294921875) [pointerCoord 1] x (expected: 365.0 actual: 364.7294921875) [pointerCoord 2] x (expected: 809.0 actual: 808.7294921875) , [pointerCoord 0] x (expected: 466.0 actual: 465.7294921875) [pointerCoord 1] x (expected: 365.0 actual: 364.7294921875) [pointerCoord 2] x (expected: 809.0 actual: 808.7294921875) , [pointerCoord 0] x (expected: 465.0 actual: 464.7294921875) [pointerCoord 1] x (expected: 364.0 actual: 363.7294921875) [pointerCoord 2] x (expected: 809.0 actual: 808.7294921875) , [pointerCoord 0] x (expected: 464.0 actual: 463.056640625) [pointerCoord 1] x (expected: 364.0 actual: 363.056640625) [pointerCoord 2] x (expected: 809.0 actual: 808.056640625) , [pointerCoord 0] x (expected: 464.0 actual: 463.056640625) [pointerCoord 1] x (expected: 364.0 actual: 363.056640625) , [pointerCoord 0] x (expected: 463.0 actual: 462.056640625) [pointerCoord 1] x (expected: 363.0 actual: 362.056640625) , [pointerCoord 0] x (expected: 461.0 actual: 458.234375) [pointerCoord 1] x (expected: 363.0 actual: 360.234375) , [pointerCoord 0] x (expected: 459.0 actual: 456.234375) [pointerCoord 1] x (expected: 363.0 actual: 360.234375) , [pointerCoord 0] x (expected: 457.0 actual: 454.234375) [pointerCoord 1] x (expected: 363.0 actual: 360.234375) , [pointerCoord 0] x (expected: 455.0 actual: 453.234375) [pointerCoord 1] x (expected: 363.0 actual: 361.234375) , [pointerCoord 0] x (expected: 453.0 actual: 449.234375) [pointerCoord 1] x (expected: 363.0 actual: 359.234375) , [pointerCoord 0] x (expected: 451.0 actual: 446.0302734375) [pointerCoord 1] x (expected: 363.0 actual: 358.0302734375) , [pointerCoord 0] x (expected: 449.0 actual: 444.0302734375) [pointerCoord 1] x (expected: 363.0 actual: 358.0302734375) , [pointerCoord 0] x (expected: 446.0 actual: 443.0302734375) [pointerCoord 1] x (expected: 363.0 actual: 360.0302734375) , [pointerCoord 0] x (expected: 446.0 actual: 443.0302734375) , [pointerCoord 0] x (expected: 443.0 actual: 440.0302734375) , [pointerCoord 0] x (expected: 440.0 actual: 433.4072265625) , [pointerCoord 0] x (expected: 436.0 actual: 431.4072265625) , [pointerCoord 0] x (expected: 432.0 actual: 427.4072265625) , [pointerCoord 0] x (expected: 428.0 actual: 417.2353515625) , [pointerCoord 0] x (expected: 424.0 actual: 413.2353515625) , [pointerCoord 0] x (expected: 419.0 actual: 412.2353515625) , [pointerCoord 0] x (expected: 415.0 actual: 408.2353515625) , [pointerCoord 0] x (expected: 409.0 actual: 398.2509765625) , [pointerCoord 0] x (expected: 404.0 actual: 393.2509765625) , [pointerCoord 0] x (expected: 398.0 actual: 372.5400390625) , [pointerCoord 0] x (expected: 392.0 actual: 371.5400390625) , [pointerCoord 0] x (expected: 384.0 actual: 363.5400390625) , [pointerCoord 0] x (expected: 374.0 actual: 345.8515625) '
```

Side-note: We should make this a presubmit test using an emulator if possible.
